### PR TITLE
feat: std.TestSetPrevRealm

### PR DESCRIPTION
**Pull request by r3v4s**
# Description

Thanks to @albttx, we now have `std.PrevRealm` #667

However it seems like doesn't work on testcase(that can be run by `gno test` command) as what we expected.
> `gno test` uses its own Context which makes few other functions (like GetRealmPath) doesn't work neither.

So I made little helper very similar to `std.TestSetOrigCaller`

---
## Discussion Need
`std.TestSetOrigCaller` takes single parameter and two different type can be passed
1. If `std.Address` type is passed, -> TestSetPrevRealm will take it as user address => return user address not realm.
2. If `string` type is passed, -> TestSetPrevRealm will take it as pkg_path => return pkg address(using bech32) and pkg_path
> Since string parameter is being used without any verification in this pr, How about reusing verification logic from here ??https://github.com/gnolang/gno/blob/408fc68d4b3c189dbc6a608c590a86c661ae232a/tm2/pkg/std/memfile.go#L33-L68

## Contributors Checklist

- [x] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests

## Maintainers Checklist

- [x] Checked that the author followed the guidelines in `CONTRIBUTING.md`
- [x] Checked the conventional-commit (especially PR title and verb, presence of `BREAKING CHANGE:` in the body)
- [x] Ensured that this PR is not a significant change or confirmed that the review/consideration process was appropriate for the change

### CODE REVIEW

**Comment by tbruyelle** [comment]
Thanks for raising this, I checked on my side and I noticed an other issue related to `PackageValue.IsRealm()` method. If we fix it, maybe we won't need this extra `std.TestSetPrevRealm` function.

`PackageValue.IsRealm()` is used in `std.PrevRealm` to determine if a frame is a realm, and if not, the frame is ignored. 

`PackageValue.IsRealm()` calls `IsRealmPath()` under the hood, which has the following body : 

```go
func IsRealmPath(pkgPath string) bool {
	// TODO: make it more distinct to distinguish from normal paths.
	if strings.HasPrefix(pkgPath, "gno.land/r/") {
		return true
	} else {
		return false
	}
}
```

The problem is that realms under `_test.gno` file have a realm path equals to `examples/gno.land/r/...` so this function returns always false, which breaks the behavior of `std.PrevRealm()`.

I'll try to provide a fix and see if that solves also the problem you raised.

**Comment by r3v4s** [comment]
@tbruyelle Thanks for response. I found other issue in this pr maybe related your `IsRealm`.

If I pass weird string to `std.TestSetPrevRealm("this_is_strange_string")` it breaks.

Would be nice if your `fix` does fix every issue on our hand, but if it doesn't I might be have to use `IsRealm` in `PrevRealm` to validate it.

**Comment by tbruyelle** [comment]
@r3v4s I was able to fix the issue I had, can you compile `gno` from this PR #896 and see if you still  have the issue on your side ? THx!

File: gnovm/tests/imports.go
Comment (zivkovicmilos) [inline gnovm/tests/imports.go:572]: Why not have this be a string input instead of any?

File: gnovm/tests/imports.go
Comment (zivkovicmilos) [inline gnovm/tests/imports.go:581]: Typo here, should probably be:
`as a caller`

File: gnovm/tests/imports.go
Comment (zivkovicmilos) [inline gnovm/tests/imports.go:612]: Typo: `invalid type`

File: examples/gno.land/r/demo/inner/inner_test.gno
Comment (zivkovicmilos) [inline examples/gno.land/r/demo/inner/inner_test.gno:26]: What do you think about naming this as `assertEqual`, or `assertE` to follow convention from other `.gno` assertion checks?
This should probably be in a package of its own, but it is out of the scope of this PR

File: gnovm/tests/imports.go
Comment (zivkovicmilos) [inline gnovm/tests/imports.go:595]: Just curious, what is this situation where `i != 1`?

File: gnovm/tests/imports.go
Comment (zivkovicmilos) [inline gnovm/tests/imports.go:597]: Do we want this kind of non-standard package name, that is defined here? cc @moul

**Review by zivkovicmilos — APPROVED** [review]
Looks good 💯 

I've left some comments that should be addressed, and pinged a few people to take another look.

Please check out @tbruyelle's [recent comment](https://github.com/gnolang/gno/pull/891#issuecomment-1591461956) 🙏

File: gnovm/tests/imports.go
Comment (r3v4s) [reply gnovm/tests/imports.go:572]: for parameter type, both `std.Address` and `string` can be used.
- if passed type is `string`, it will set PrevRealm to a `Realm.
- if passed type is `std.Address`, it will set PrevRealm to a (normal) user

if `TestSetPrevRealm` takes only string type, we have to determine whether that value is user address or pkg_path, which we need clear way to determine it.

File: gnovm/tests/imports.go
Comment (r3v4s) [reply gnovm/tests/imports.go:597]: I wanted pkg_path value used on that line should never exist on chain. If it exists maybe there will be some side effect ( or maybe I'm worrying too much :D )

File: gnovm/tests/imports.go
Comment (r3v4s) [reply gnovm/tests/imports.go:595]: While testing with `gno test` command, it returned `m.NumFrames() = 6` 

https://github.com/gnolang/gno/blob/f50f33de653bd660fe1a2ceae430fd7885e97e9c/gnovm/stdlibs/stdlibs.go#L318-L335


`PrevRealm` iterates all over `m.NumFrames()` and look for `fr.LastPackage`changes.
To satisfy changes, `NumFrame[i] // i != 1` needs dummy value

File: gnovm/tests/imports.go
Comment (ajnavarro) [reply gnovm/tests/imports.go:572]: I would prefer to have specific methods for specific cases, like `TestSetPRevRealm` and `TestSetPrevUser`, instead of using `any` making it more difficult for the user to figure out what that method accepts.

Comment (r3v4s) [reply gnovm/tests/imports.go:572]: > 

good point, I really haven't thought about dividing these kind of functions. I'll divide it

**Comment by r3v4s** [comment]
Just removed logic for setting caller to PrevRealm, looks like `st.TestSetOrigCaller()` does that

**Comment by r3v4s** [comment]
- `TestSetPrevAddr` removes previous pkg path, and sets address (just like `user` called it directly)

- `TestSetPrevRealm` keeps origin caller, and sets previous pkg path

**Comment by r3v4s** [comment]
update: remove nested testcase, and made simple one in `gno.land/r/demo/test`

**Review by moul — APPROVED** [review]

**Comment by tbruyelle** [comment]
My take on this PR stays the same, considering #1048, we don't need those extra `TestSetPrev*` functions. But since it's blocked, do what you think is best.

**Review by piux2 — DISMISSED** [review]
We shouldn't need to inject a TestSetPrevRealm as a native method for testing.

We will need a separate RFC for the reasons outlined below:

Given the numerous incidents on other chains where millions were mistakenly sent to contract addresses instead of user addresses, it's crucial that we distinguish between contract entities (represented by package paths) and EOA (Externally Owned Account) entities (represented by addresses) in gno contract.

- Instead of returning an address for PrevRealm, we should return the realm package.

- Furthermore, we need to address the ambiguity between returning user account addresses and realm identity from PrevRealm. We might want to explicitly check the returned value to determine whether we are validating an EOA or a contract for entry method we try guard. 

- Not sure we have enough strong use cases where we had to pass contract addresses instead of realm package paths between contracts in gno.

**Comment by r3v4s** [comment]
here is short briefing for why I implemented it.

let's say... certain contact want to receive grc20 token from user(wallet, address).
to test that logic with test case, something like this will be used.

```go
# gno.mod ( gno.land/r/demo/receive )
# contract_test.gno

import (
	"std"
	"testing"
	foo "gno.land/r/demo/foo20"
)


const sender = std.Address("g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5")


func TestFunc(t *testing.T) {
	std.TestSetOrigCaller(sender)

	foo.Transfer(std.GetOrigPkgAddr(), "12345")
}
```

because `foo` uses `caller` address from `std.PrevRealm().Addr()`
https://github.com/gnolang/gno/blob/dce345f31b8fc6be3e831778983345d36f01e31f/examples/gno.land/r/demo/foo20/foo20.gno#L51

actual caller will be same with `std.DerivePkgAddr("gno.land/r/demo/receive")` rather than my expected value `g1jg8mtutu9khhfwc4nxmuhcpftf0pajdhfvsqf5`

**Comment by thehowl** [comment]
@piux2 

> * Instead of returning an address for PrevRealm, we should return the realm package.

Mhh, when we return from PrevRealm, calls from other realms have a PkgPath, if they are users they don't. The helper `IsUser()` is also provided to help test this.

Does this not address all of your concerns?

**Comment by thehowl** [comment]
So, I did a second review and agreed that the approach here is not correct.

- Instead of overwriting the previous frames, we should add a new frame that marks the PrevRealm
- When setting a PrevRealm, the address should still be set.
- The current approach pollutes the frame stack, presumably because of a missing return (so fixable), but I think a different approach in all is warranted.

I have been discussing about this with @albttx on how to best do this and he's working on a PR to iterate and improve on your work, @r3v4s. I also have some ideas on a different API we could have for the `std.TestSetPrevRealm` functionality; I'll make a separate issue for it and add it to our agenda for today's call.

----

For @piux2's concerns,

> We shouldn't need to inject a TestSetPrevRealm as a native method for testing.

We should, (or at least have a similar mechanism), otherwise the only way to test a realm when being called from other realms is through filetests, which I personally think is inflexible (ie. can't create table tests, trying to call from different realms) and creates pollution in a realm's directory as a realm grows in complexity.

> We will need a separate RFC for the reasons outlined below:
> 
> Given the numerous incidents on other chains where millions were mistakenly sent to contract addresses instead of user addresses, it's crucial that we distinguish between contract entities (represented by package paths) and EOA (Externally Owned Account) entities (represented by addresses) in gno contract.

This is still one of the design goals in the implementation of `PrevRealm` and it can be easily done. On `std.Realm`, you can determine whether the caller (or previous-realm, in the paradigm proposed by Jae whereby even users can be considered "realms"), is an end user or a published gno realm by calling the method IsUser(). More specifically, this works by testing `return r.pkgPath == ""` -- ie. end-users don't have a pkgPath, published realms do.

>     * Instead of returning an address for PrevRealm, we should return the realm package.

PrevRealms should have addresses, as realms have addresses with funds, and a caller may be looking for that information.

**Comment by tbruyelle** [comment]
> My take on this PR stays the same, considering #1048, we don't need those extra `TestSetPrev*` functions. But since #1048 is blocked, do what you think is best.

So I though I fixed that last week but I probably forgot to click the button...

This comment from me mentions the wrong PR, I wanted to mention this PR #896, which is merged since a long time ago. IMO, this PR mitigates the needs of `TestSetPrevRealm` because it ensures the _test files have a realm (which wasn't the case before).

The way this works is the following, if there's a gno.mod file, the realm is the one from the gno.mod, so the _test shares the same realm as the tested realm. If there's no gno.mod file, the realm of the _test is generated randomly, so it's different from the tested realm. Looks weird when read like that but anyway I don't think _test files should be used to test `PrevRealm` situration.

_filtest on the other hand are goods for PrevRealm testing, and for them the solution exists from the beginning, you just need to add a `// PKGPATH:` instruction to define the realm.

**Comment by tbruyelle** [comment]
> Instead of overwriting the previous frames, we should add a new frame that marks the PrevRealm

I don't think it's a good idea to manually manipulate the frames. The thing is complex enough, let's not add other way to confuse ourselves.

**Comment by thehowl** [comment]
@tbruyelle my comment goes in a direction of removing filetests which I kind of had subconsciously but hadn't written/talked about it yet :stuck_out_tongue: . I've created an issue so we can discuss that and understand if it is a common feeling / my proposal makes sense: https://github.com/gnolang/gno/issues/1346

**Comment by r3v4s** [comment]
After #764 merged, mocking(or overwriting) codes got really short (only few lines)

In my local testing, these `TestSet` is really helpful for situation like I've mentioned [above](https://github.com/gnolang/gno/pull/891#issuecomment-1759351312).

**Review by leohhhn — APPROVED** [review]
Revisiting this - We should definitely have `TestSetPrevRealm`, just like we do with other  ctx variables. 

@r3v4s, can you please write txtar tests for this, and possibly simplify them to not use a separate package)? 🙏

**Comment by r3v4s** [comment]
@leohhhn I don't think with `txtar testing` I can call `Test*` functions. Aren't they meant to only callable from `gno test` ??

**Comment by zivkovicmilos** [comment]
Merged `master` locally and all tests are passing 💯

**Pull request merged**
